server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -987,8 +972,27 @@ private Engine.IndexResult applyIndexOperation(
         long autoGeneratedTimeStamp,
         boolean isRetry,
         Engine.Operation.Origin origin,
-        SourceToParse sourceToParse
+        SourceToParse sourceToParse,
+        String id
     ) throws IOException {
+        if (indexSettings.isSegRepEnabled() && routingEntry().primary() == false && getReplicationEngine().isPresent()) {
```

Comment:
nit - There is no need to validate the engine type here once checking the routing & index setting.